### PR TITLE
CLN: core/dtypes/cast.py::maybe_downcast_to_dtype

### PR DESCRIPTION
**Pull request by arw2019**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -159,29 +168,19 @@ def maybe_downcast_to_dtype(result, dtype):
         return converted
 
     # a datetimelike
-    # GH12821, iNaT is casted to float
-    if dtype.kind in ["M", "m"] and result.dtype.kind in ["i", "f"]:
-        if hasattr(dtype, "tz"):
+    # GH12821, iNaT is cast to float
+    if is_datetime_or_timedelta_any_dtype(dtype) and result.dtype.kind in ["i", "f"]:
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:172]: the dtype.kind checks are much more performant

Comment (jreback) [reply pandas/core/dtypes/cast.py:172]: i agree, just leave this as is i think for L163 (and can remove the comment which isold)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py]: Ok! Done both

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -159,29 +167,19 @@ def maybe_downcast_to_dtype(result, dtype):
         return converted
 
     # a datetimelike
-    # GH12821, iNaT is casted to float
+    # GH12821, iNaT is cast to float
     if dtype.kind in ["M", "m"] and result.dtype.kind in ["i", "f"]:
-        if hasattr(dtype, "tz"):
+
+        if is_datetime_or_timedelta_dtype(dtype):
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:173]: i think this code here can be greatly simplified, see what the incoming inputs are and i bet you can adjust it

Comment (jreback) [reply pandas/core/dtypes/cast.py:173]: IOW the way this is written L173 is *always* hit

Comment (arw2019) [reply pandas/core/dtypes/cast.py]: tried this. There are three tests that use the else block:
```
FAILED pandas/tests/dtypes/cast/test_downcast.py::test_datetime_with_timezone[True]
FAILED pandas/tests/frame/methods/test_combine_first.py::TestDataFrameCombineFirst::test_combine_first_timezone
FAILED pandas/tests/groupby/test_function.py::test_arg_passthru - TypeError: ...
```
Reverting

Comment (jreback) [reply pandas/core/dtypes/cast.py:173]: and you have L177. so i would revert to the original code here as well (except for the period change);

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/cast.py (hunk 2 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py]: Done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @arw2019

**Comment by simonjayhawkins** [comment]
@arw2019 this solution could be to revert this and re-open? The mypy errors are

```
pandas\core\dtypes\cast.py:173: error: Item "type" of "Union[ExtensionDtype, Any, Type[object]]" has no attribute "type"  [union-attr]
pandas\core\dtypes\cast.py:179: error: Item "ExtensionDtype" of "Union[ExtensionDtype, Any, Type[object]]" has no attribute "freq"  [union-attr]
pandas\core\dtypes\cast.py:179: error: Item "type" of "Union[ExtensionDtype, Any, Type[object]]" has no attribute "freq"  [union-attr]
```

so i'm a bit reluctant to ignore to get mypy to green.

The issue is related to https://github.com/pandas-dev/pandas/pull/37024#discussion_r503495620

so maybe only str and dtypeObj are accepted. so could also change to Union[str, DtypeObj] but this I _think_ is unusual and should normally be DtypeObj or Dtype.

**Comment by simonjayhawkins** [comment]
we'll go with the revert for now even though the fix is straightfoward to allow unpressured discussion on the topic. OK with reopening this PR?

**Pull request merged**